processor/src/range/aux_trace.rs

```diff
@@ -0,0 +1,235 @@
+    ) -> (Vec<E>, Vec<E>) {
+        // compute the inverses for range checks performed by operations.
+        let (_, inv_row_values) =
+            build_lookup_table_row_values(&self.cycle_range_check_values(), alphas);
+
+        // allocate memory for the running product column and set the initial value to ONE
+        let mut q = unsafe { uninit_vector(main_trace.num_rows()) };
+        let mut p1 = unsafe { uninit_vector(main_trace.num_rows()) };
+        q[0] = E::ONE;
+        p1[0] = E::ONE;
+
+        // keep track of the last updated row in the `p1` running product column. the `q` column
+        // index is always one row behind, since `q` is filled with intermediate values in the same
+        // row as the operation is executed, whereas `p1` is filled with result values that are
+        // added to the next row after the operation's execution.
+        let mut p1_idx = 0;
+        // keep track of the next row to be included from the user op range check values.
+        let mut rc_user_op_idx = 0;
+
+        // the first half of the trace only includes values from the operations.
+        for (clk, range_checks) in self
+            .cycle_range_checks
+            .iter()
+            .take_while(|(&step, _)| step <= self.start_16bit)
+        {
+            let clk = *clk;
+
+            // if we skipped some cycles since the last update was processed, values in the last
+            // updated row should by copied over until the current cycle.
+            if p1_idx < clk {
+                let last_value = p1[p1_idx];
+                p1[(p1_idx + 1)..=clk].fill(last_value);
+                q[p1_idx..clk].fill(E::ONE);
+            }
+
+            // move the column pointers to the next row.
+            p1_idx = clk + 1;
+
+            // update the intermediate values in the q column.
+            q[clk] = range_checks.to_stack_value(alphas);
+
+            // include the operation lookups in the running product.
+            p1[p1_idx] = p1[clk] * inv_row_values[rc_user_op_idx];
+            rc_user_op_idx += 1;
+        }
+
+        // if we skipped some cycles since the last update was processed, values in the last
+        // updated row should by copied over until the current cycle.
+        if p1_idx < self.start_16bit {
+            let last_value = p1[p1_idx];
+            p1[(p1_idx + 1)..=self.start_16bit].fill(last_value);
+            q[p1_idx..self.start_16bit].fill(E::ONE);
+        }
+
+        // for the 16-bit section of the range checker table, include `z` in the running product at
+        // each step and remove lookups from user ops at any step where user ops were executed.
+        for (row_idx, (hint, lookup)) in self
+            .row_flags
+            .iter()
+            .zip(main_trace.get_column(V_COL_IDX).iter())
+            .enumerate()
+            .take(main_trace.num_rows() - NUM_RAND_ROWS - 1)
+            .skip(self.start_16bit)
+        {
+            let last_value = p1[row_idx];
```

Comment:
minor nit: doesn't seem like `last_value` is used anywhere besides line 207. Maybe just replace it there with `p1[row_idx]`?

Reply:
Yep, updated